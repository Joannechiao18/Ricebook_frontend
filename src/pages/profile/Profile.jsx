import React, { useState, useEffect } from "react";

import { Link, Navigate, useNavigate } from "react-router-dom";
import PlaceIcon from "@mui/icons-material/Place";
import EmailOutlinedIcon from "@mui/icons-material/EmailOutlined";
import CakeIcon from "@mui/icons-material/Cake";
import PhoneIcon from "@mui/icons-material/Phone";
import ArrowBackIosIcon from "@mui/icons-material/ArrowBackIos";
import EditProfileModal from "../../components/modal/EditProfileModal";
import "./profile.scss";
import AddAPhotoIcon from "@mui/icons-material/AddAPhoto";
import { useSelector, useDispatch } from "react-redux";
import { selectUser } from "../../reducer/authReducer";
import { login } from "../../actions/authActions";
import styled from "styled-components";

const BaseButton = styled.button`
  width: auto;
  border: none;
  border-radius: 5px;
  padding: 5px 15px;
  font-size: 13px;
  cursor: pointer;
  transition: background-color 0.2s ease, color 0.2s ease;
  font-weight: 600;
  font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto,
    "Helvetica Neue", Arial, "Noto Sans", sans-serif, "Apple Color Emoji",
    "Segoe UI Emoji", "Segoe UI Symbol", "Noto Color Emoji";
`;

const EditButton = styled(BaseButton)`
  color: white;
  background-color: #938eef; // Mimicking Bootstrap btn-danger color
  margin-right: 5px; // Add some margin to the right of the Cancel button

  &:hover {
    background-color: #7a75d6; // Darker shade for hover effect
  }
`;

const LinkButton = styled(BaseButton)`
  color: white;
  background-color: #3b5998; // Color for Link Accounting button
  margin-right: 5px;

  &:hover {
    background-color: #324c85;
  }
`;

const UnlinkButton = styled(LinkButton)`
  width: auto;
  background-color: #d36c5c; // Color for Unlink Accounting button

  &:hover {
    background-color: #a22b2b;
  }
`;

const Profile = () => {
  const currentUser = useSelector(selectUser) || {};

  const dispatch = useDispatch();
  const [isModalOpen, setIsModalOpen] = useState(false);
  const [profilePicture, setProfilePicture] = useState(null);

  // Add state variables for linking/unlinking accounting
  const [isAccountingLinked, setIsAccountingLinked] = useState(false);
  const [isLinkingInProgress, setIsLinkingInProgress] = useState(false);
  const navigate = useNavigate();

  const [profileData, setProfileData] = useState({
    username: "",
    email: "",
    dob: "",
    phone: "",
    zipcode: "",
    password: "",
    avatar: "",
  });

  useEffect(() => {
    const fetchData = async () => {
      try {
        const username = currentUser.username;

        const fetchJsonData = async (url) => {
          const response = await fetch(url, {
            method: "GET",
            credentials: "include",
            headers: {
              "Content-Type": "application/json",
            },
          });

          if (!response.ok) {
            throw new Error(
              `Error: ${response.status} - ${response.statusText}`
            );
          }

          return await response.json();
        };

        // Fetch user details (email, phone, etc.)
        const emailData = await fetchJsonData(
          //`https://yw187server-3d9494142af2.herokuapp.com/email/${username}`
          `https://ricebookserveryw187-8fbcb305db50.herokuapp.com/email/${username}`
        );

        const dobData = await fetchJsonData(
          //`https://yw187server-3d9494142af2.herokuapp.com/email/${username}`
          `https://ricebookserveryw187-8fbcb305db50.herokuapp.com/dob/${username}`
        );

        const phoneData = await fetchJsonData(
          //`https://yw187server-3d9494142af2.herokuapp.com/phone/${username}`
          `https://ricebookserveryw187-8fbcb305db50.herokuapp.com/phone/${username}`
        );
        const zipcodeData = await fetchJsonData(
          //`https://yw187server-3d9494142af2.herokuapp.com/zipcode/${username}`
          `https://ricebookserveryw187-8fbcb305db50.herokuapp.com/zipcode/${username}`
        );
        const avatarData = await fetchJsonData(
          `https://ricebookserveryw187-8fbcb305db50.herokuapp.com/avatar/${username}`
        );

        // Update state with the fetched data
        setProfileData({
          username: username,
          email: emailData.email,
          dob: dobData.dob || "",
          phone: phoneData.phone || "",
          zipcode: zipcodeData.zipcode || "",
          avatar: avatarData.avatar || "",
        });
      } catch (error) {
        console.error("Error fetching profile data:", error.message);
      }
    };

    fetchData();
  }, [currentUser]);

  const handleEditClick = () => {
    setIsModalOpen(true);
  };

  const handleModalClose = () => {
    setIsModalOpen(false);
  };

  const handleUpdateProfile = async (formData) => {
    const updatedProfileData = {
      username: formData.username,
      email: formData.email,
      phone: formData.phone,
      password: formData.password,
      zipcode: formData.zipcode,
    };

    try {
      // Make API requests to update user information
      await Promise.all([
        // Update zipcode
        fetch(
          "https://ricebookserveryw187-8fbcb305db50.herokuapp.com/zipcode",
          {
            method: "PUT",
            credentials: "include",
            headers: {
              "Content-Type": "application/json",
            },
            body: JSON.stringify({
              zipcode: updatedProfileData.zipcode,
            }),
          }
        ),

        // Update email
        fetch("https://ricebookserveryw187-8fbcb305db50.herokuapp.com/email", {
          method: "PUT",
          credentials: "include",
          headers: {
            "Content-Type": "application/json",
          },
          body: JSON.stringify({
            email: updatedProfileData.email,
          }),
        }),

        // Update phone
        fetch("https://ricebookserveryw187-8fbcb305db50.herokuapp.com/phone", {
          method: "PUT",
          credentials: "include",
          headers: {
            "Content-Type": "application/json",
          },
          body: JSON.stringify({
            phone: updatedProfileData.phone,
          }),
        }),
      ]);

      // Create a new object with the updated profile data
      const updatedUser = {
        ...currentUser,
        ...updatedProfileData,
      };

      // Dispatch the updated profile data to Redux
      dispatch(login(updatedUser));

      setIsModalOpen(false);
    } catch (error) {
      console.error("Error updating profile:", error);
      // Handle error and show a notification to the user
    }
  };

  const handleFileChange = (event) => {
    const file = event.target.files[0];
    if (file) {
      const reader = new FileReader();

      reader.onloadend = () => {
        setProfilePicture(reader.result);
        handleUploadProfilePicture(file); // Call upload function here
      };

      reader.readAsDataURL(file);
    }
  };

  const handleUploadProfilePicture = async (file) => {
    const formData = new FormData();
    formData.append("image", file);

    if (file) {
      const response = await fetch(
        `https://ricebookserveryw187-8fbcb305db50.herokuapp.com/avatar`,
        {
          method: "PUT",
          credentials: "include",

          body: formData,
        }
      );

      //傳第二次responese才會是ok問題

      if (!response.ok) {
        throw new Error("Failed to update avatar.");
      }

      const updatedData = await response.json();

      dispatch(
        login({
          ...currentUser,
          avatar: updatedData.avatar,
        })
      );

      setProfileData((prevData) => ({
        ...prevData,
        avatar: updatedData.avatar,
      }));
    }
  };

  // Function to handle linking accounting
  const handleLinkAccounting = async () => {
    setIsLinkingInProgress(true);

    try {
      // Replace this with your actual linking logic
      // Example: const response = await fetch("link_accounting_url", { method: "POST" });

      // If linking is successful, update the state
      setIsAccountingLinked(true);
    } catch (error) {
      console.error("Error linking accounting:", error);
    } finally {
      setIsLinkingInProgress(false);
    }
  };

  const handleUnlinkAccounting = async () => {
    console.log("unlink");
    setIsLinkingInProgress(true);

    try {
      const response = await fetch(
        "https://ricebookserveryw187-8fbcb305db50.herokuapp.com/unlinkThirdPartyUser",
        {
          method: "DELETE",
          credentials: "include",
          headers: {
            "Content-Type": "application/json",
          },
          body: JSON.stringify({
            username: currentUser.username,
            isThirdPartyUser: currentUser.isThirdPartyUser,
          }),
        }
      );

      console.log(response);

      if (!response.ok) {
        throw new Error(`Failed to delete account: ${response.statusText}`);
      }

      // Handle successful account deletion
      // Redirect to the login page
      //navigate("/login");
      window.location.href =
        "https://ricebookserveryw187-8fbcb305db50.herokuapp.com/auth/google/callback";
    } catch (error) {
      console.error("Error deleting account:", error);
      // Handle error and show a notification to the user
    } finally {
      setIsLinkingInProgress(false);
    }
  };

  useEffect(() => {}, [profilePicture]);

  return (
    <div className="profile container mt-4">
      <Link to="/">
        <ArrowBackIosIcon className="back-button" />
      </Link>
      <div className="row mt-5">
        <div className="col-12 text-center mt-3">
          <div className="profile-pic-wrapper">
            <img
              src={profileData.avatar}
              //alt={currentUser ? profileData.username : "User"}
              className="img-fluid rounded-circle profile-pic"
            />

            <input
              type="file"
              id="profile-pic-input"
              style={{ display: "none" }}
              onChange={(e) => handleFileChange(e, handleUploadProfilePicture)}
              accept="image/*"
            />
            <label
              htmlFor="profile-pic-input"
              className="profile-picture-label"
            >
              <AddAPhotoIcon
                onClick={handleUploadProfilePicture}
                style={{ fontSize: 18, color: "#938eef" }}
              />
            </label>
          </div>
        </div>
      </div>
      <div className="row">
        <div className="col-4 text-center"></div>
        <div className="col-4 text-center">
          <h3>
            <strong>{profileData.username}</strong>
          </h3>
          <p>
            <EmailOutlinedIcon /> {profileData.email}
          </p>
          <p>
            <CakeIcon />
            {profileData.dob}
          </p>
          <p>
            <PhoneIcon /> {profileData.phone}
          </p>
          <p>
            <PlaceIcon />
            {profileData.zipcode}
          </p>
          {/*<p>
            <VisibilityOffIcon />
            {profileData.password && "*".repeat(profileData.password.length)}
            </p>*/}
          <div>
            <EditButton onClick={handleEditClick}>Edit</EditButton>
          </div>
          {/* Place the Link and Unlink buttons in the same line */}
          <div
            className="d-flex justify-content-between align-items-center"
            style={{ marginTop: "10px" }}
          >
            <div>
              <LinkButton
                onClick={handleLinkAccounting}
                disabled={isLinkingInProgress}
              >
                {isLinkingInProgress ? "Linking..." : "Link Accounting"}
              </LinkButton>
            </div>
            <div>
              <UnlinkButton
                onClick={handleUnlinkAccounting}
                disabled={isLinkingInProgress}
              >
                {isLinkingInProgress ? "Unlinking..." : "Unlink Accounting"}
              </UnlinkButton>
            </div>
          </div>
        </div>
      </div>
      <EditProfileModal
        isOpen={isModalOpen}
        onRequestClose={handleModalClose}
        user={currentUser}
        onUpdate={handleUpdateProfile}
      />
    </div>
  );
};

export default Profile;
