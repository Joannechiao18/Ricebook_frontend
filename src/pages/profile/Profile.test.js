import React from "react";
import { render, screen } from "@testing-library/react";
import { Provider } from "react-redux";
import { BrowserRouter as Router } from "react-router-dom";
import configureStore from "redux-mock-store";
import Profile from "../profile/Profile";

// Import the EditProfileModalMock component
//import EditProfileModalMock from "../profile/EditProfileModal.mock";

// Create a mock Redux store
const mockStore = configureStore([]);
const store = mockStore({
  auth: {
    currentUser: {
      username: "testuser",
      email: "test@example.com",
      phone: "1234567890",
      zipcode: "12345",
      password: "password",
      profilePic: "https://via.placeholder.com/150",
    },
  },
});

jest.mock("../../components/modal/EditProfileModal", () => {
  // Directly return the mock component here
  return () => <div>EditProfileModalMock</div>;
});

describe("Profile Component", () => {
  it("renders the logged-in user's username", () => {
    render(
      <Provider store={store}>
        <Router>
          <Profile />
        </Router>
      </Provider>
    );

    const usernameElement = screen.getByText("testuser");

    expect(usernameElement).toBeInTheDocument();
  });

  it("matches snapshot", () => {
    const { asFragment } = render(
      <Provider store={store}>
        <Router>
          <Profile />
        </Router>
      </Provider>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
