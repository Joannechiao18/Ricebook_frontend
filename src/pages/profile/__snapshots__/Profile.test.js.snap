// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile Component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="profile container mt-4"
  >
    <a
      href="/"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium back-button css-i4bv87-MuiSvgIcon-root"
        data-testid="ArrowBackIosIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M11.67 3.87 9.9 2.1 0 12l9.9 9.9 1.77-1.77L3.54 12z"
        />
      </svg>
    </a>
    <div
      class="row mt-5"
    >
      <div
        class="col-12 text-center mt-3"
      >
        <div
          class="profile-pic-wrapper"
        >
          <img
            alt="testuser"
            class="img-fluid rounded-circle profile-pic"
            src="https://via.placeholder.com/150"
          />
          <input
            accept="image/*"
            id="profile-pic-input"
            style="display: none;"
            type="file"
          />
          <label
            class="profile-picture-label"
            for="profile-pic-input"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="AddAPhotoIcon"
              focusable="false"
              style="font-size: 18px; color: rgb(147, 142, 239);"
              viewBox="0 0 24 24"
            >
              <path
                d="M3 4V1h2v3h3v2H5v3H3V6H0V4h3zm3 6V7h3V4h7l1.83 2H21c1.1 0 2 .9 2 2v12c0 1.1-.9 2-2 2H5c-1.1 0-2-.9-2-2V10h3zm7 9c2.76 0 5-2.24 5-5s-2.24-5-5-5-5 2.24-5 5 2.24 5 5 5zm-3.2-5c0 1.77 1.43 3.2 3.2 3.2s3.2-1.43 3.2-3.2-1.43-3.2-3.2-3.2-3.2 1.43-3.2 3.2z"
              />
            </svg>
          </label>
        </div>
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="col-4 text-center"
      />
      <div
        class="col-4 text-center"
      >
        <h3>
          <strong>
            testuser
          </strong>
        </h3>
        <p>
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="EmailOutlinedIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M22 6c0-1.1-.9-2-2-2H4c-1.1 0-2 .9-2 2v12c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V6zm-2 0-8 5-8-5h16zm0 12H4V8l8 5 8-5v10z"
            />
          </svg>
           test@example.com
        </p>
        <p>
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="PhoneIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M6.62 10.79c1.44 2.83 3.76 5.14 6.59 6.59l2.2-2.2c.27-.27.67-.36 1.02-.24 1.12.37 2.33.57 3.57.57.55 0 1 .45 1 1V20c0 .55-.45 1-1 1-9.39 0-17-7.61-17-17 0-.55.45-1 1-1h3.5c.55 0 1 .45 1 1 0 1.25.2 2.45.57 3.57.11.35.03.74-.25 1.02l-2.2 2.2z"
            />
          </svg>
           1234567890
        </p>
        <p>
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="PlaceIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 2C8.13 2 5 5.13 5 9c0 5.25 7 13 7 13s7-7.75 7-13c0-3.87-3.13-7-7-7zm0 9.5c-1.38 0-2.5-1.12-2.5-2.5s1.12-2.5 2.5-2.5 2.5 1.12 2.5 2.5-1.12 2.5-2.5 2.5z"
            />
          </svg>
          12345
        </p>
        <p>
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="VisibilityOffIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 7c2.76 0 5 2.24 5 5 0 .65-.13 1.26-.36 1.83l2.92 2.92c1.51-1.26 2.7-2.89 3.43-4.75-1.73-4.39-6-7.5-11-7.5-1.4 0-2.74.25-3.98.7l2.16 2.16C10.74 7.13 11.35 7 12 7zM2 4.27l2.28 2.28.46.46C3.08 8.3 1.78 10.02 1 12c1.73 4.39 6 7.5 11 7.5 1.55 0 3.03-.3 4.38-.84l.42.42L19.73 22 21 20.73 3.27 3 2 4.27zM7.53 9.8l1.55 1.55c-.05.21-.08.43-.08.65 0 1.66 1.34 3 3 3 .22 0 .44-.03.65-.08l1.55 1.55c-.67.33-1.41.53-2.2.53-2.76 0-5-2.24-5-5 0-.79.2-1.53.53-2.2zm4.31-.78 3.15 3.15.02-.16c0-1.66-1.34-3-3-3l-.17.01z"
            />
          </svg>
          ********
        </p>
        <button
          class="sc-beyTiQ sc-guDMob kXHBDX hsoiRu"
        >
          Edit
        </button>
      </div>
      <div
        class="col-4 text-center"
      />
    </div>
    <div>
      EditProfileModalMock
    </div>
  </div>
</DocumentFragment>
`;
